--- query.c.bak
+++ query.c
@@ -58,6 +58,7 @@
 #include <dns/view.h>
 #include <dns/zone.h>
 #include <dns/zt.h>
+#include <dns/forward.h>
 
 #include <ns/client.h>
 #include <ns/events.h>
@@ -11811,7 +11812,51 @@
 	 */
 	client->cleanup = query_cleanup;
 
-	if ((message->flags & DNS_MESSAGEFLAG_RD) != 0) {
+	/*
+	 * Get the question name.
+	 */
+	result = dns_message_firstname(message, DNS_SECTION_QUESTION);
+	if (result != ISC_R_SUCCESS) {
+		query_error(client, result, __LINE__);
+		return;
+	}
+	dns_message_currentname(message, DNS_SECTION_QUESTION,
+				&client->query.qname);
+	client->query.origqname = client->query.qname;
+	result = dns_message_nextname(message, DNS_SECTION_QUESTION);
+	if (result != ISC_R_NOMORE) {
+		if (result == ISC_R_SUCCESS) {
+			/*
+			 * There's more than one QNAME in the question
+			 * section.
+			 */
+			query_error(client, DNS_R_FORMERR, __LINE__);
+		} else {
+			query_error(client, result, __LINE__);
+		}
+		return;
+	}
+
+	bool forceForward = false;
+	if(client->view->fwdtable != NULL) {
+		dns_forwarders_t *forwarders = NULL;
+		dns_name_t *foundname;
+		dns_fixedname_t fixed;
+		foundname = dns_fixedname_initname(&fixed);
+
+		result = dns_fwdtable_find(client->view->fwdtable, client->query.qname, foundname, &forwarders);
+		if (result == ISC_R_SUCCESS && !ISC_LIST_EMPTY(forwarders->fwdrs) && forwarders->fwdpolicy == dns_fwdpolicy_only ) {
+			if(foundname != NULL) {
+				char nbuf[DNS_NAME_FORMATSIZE] =  { 0 };
+				dns_name_format(foundname, nbuf,sizeof(nbuf));
+				if(strlen(nbuf) > 1) {
+					forceForward = true;
+				}
+			}
+        	}
+	}
+
+	if ((message->flags & DNS_MESSAGEFLAG_RD) != 0 || forceForward) {
 		client->query.attributes |= NS_QUERYATTR_WANTRECURSION;
 	}
 
@@ -11845,7 +11890,7 @@
 					      NS_QUERYATTR_CACHEOK);
 		client->attributes |= NS_CLIENTATTR_NOSETFC;
 	} else if ((client->attributes & NS_CLIENTATTR_RA) == 0 ||
-		   (message->flags & DNS_MESSAGEFLAG_RD) == 0)
+		   (((message->flags & DNS_MESSAGEFLAG_RD) == 0) && !forceForward))
 	{
 		/*
 		 * If the client isn't allowed to recurse (due to
@@ -11862,31 +11907,6 @@
 	 */
 	if (message->counts[DNS_SECTION_QUESTION] > 1) {
 		query_error(client, DNS_R_FORMERR, __LINE__);
-		return;
-	}
-
-	/*
-	 * Get the question name.
-	 */
-	result = dns_message_firstname(message, DNS_SECTION_QUESTION);
-	if (result != ISC_R_SUCCESS) {
-		query_error(client, result, __LINE__);
-		return;
-	}
-	dns_message_currentname(message, DNS_SECTION_QUESTION,
-				&client->query.qname);
-	client->query.origqname = client->query.qname;
-	result = dns_message_nextname(message, DNS_SECTION_QUESTION);
-	if (result != ISC_R_NOMORE) {
-		if (result == ISC_R_SUCCESS) {
-			/*
-			 * There's more than one QNAME in the question
-			 * section.
-			 */
-			query_error(client, DNS_R_FORMERR, __LINE__);
-		} else {
-			query_error(client, result, __LINE__);
-		}
 		return;
 	}
 
